File: d_fourth_lesson/src/fibonacci_server.cpp
#include <ros/ros.h>
#include <actionlib/server/simple_action_server.h>
#include "d_fourth_lesson/FibonacciAction.h"

/**
 * @brief FibonacciAction -> gives you the Action itself
 * @brief FibonacciFeedback -> gives you the Action feedback
 * @brief FibonacciResult -> gives you the Action result
 *
 */

class FibonacciAction
{
public:
    // action_server and action_name are declared in the protected variable
    FibonacciAction(std::string name) : action_server(name, boost::bind(&FibonacciAction::executeCB, this, boost::placeholders::_1), false),
                                        action_name(name)
    {
        action_server.start();
    }

    ~FibonacciAction(void) {}

    void executeCB(const d_fourth_lesson::FibonacciGoalConstPtr &goal)
    {
        // helper variables
        ros::Rate rate(1);
        bool success = true;

        // push_back the seeds for the fibonacci sequence
        feedback.sequence.clear();
        feedback.sequence.push_back(0);
        feedback.sequence.push_back(1);

        // publish info to the console for the user
        ROS_INFO(
            "%s: Executing, creating fibonacci sequence of order %i with seeds %i, %i",
            action_name.c_str(),
            goal->order,
            feedback.sequence[0],
            feedback.sequence[1]);

        // start executing the action
        for (int i = 1; i <= goal->order; i++)
        {
            // check that preempt has not been requested by the client
            if (action_server.isPreemptRequested() || !ros::ok())
            {
                ROS_INFO("%s: Preempted", action_name.c_str());

                // set the action state to preempted
                action_server.setPreempted();
                success = false;
                break;
            }

            feedback.sequence.push_back(feedback.sequence[i] + feedback.sequence[i - 1]);
            action_server.publishFeedback(feedback);

            // this sleep is not necessary, the sequence is computed at 1 Hz for demonstration purposes
            rate.sleep();

            if (success)
            {
                result.sequence = feedback.sequence;
                ROS_INFO("%s: Succeeded", action_name.c_str());

                // set the action state to succeeded
                action_server.setSucceeded(result);
            }
        }
    }

protected:
    ros::NodeHandle nh;
    actionlib::SimpleActionServer<d_fourth_lesson::FibonacciAction> action_server;

    std::string action_name;

    d_fourth_lesson::FibonacciFeedback feedback;
    d_fourth_lesson::FibonacciResult result;
};

int main(int argc, char **argv)
{
    ros::init(argc, argv, "fibonacci");

    FibonacciAction fibonacci("fibonacci");

    ros::spin();

    return 0;
}

File: d_fourth_lesson/src/fibonacci_server.py
#! /usr/bin/python3

import rospy
import actionlib
from d_fourth_lesson.msg import FibonacciFeedback, FibonacciResult, FibonacciAction, FibonacciGoal


class FibonacciAction(object):
    # create messages that are used to publish feedback/result
    _feedback = FibonacciFeedback()
    _result = FibonacciResult()

    def __init__(self, name):
        self._action_name = name
        self._action_server = actionlib.SimpleActionServer(
            self._action_name, FibonacciAction, execute_cb=self.execute_cb, auto_start=False)
        self._action_server.start()

    def execute_cb(self, goal: FibonacciGoal):
        # helper variables
        rate = rospy.Rate(1)
        success = True

        # append the seeds for the fibonacci sequence
        self._feedback.sequence = []
        self._feedback.sequence.append(0)
        self._feedback.sequence.append(1)

        # publish info to the console for the user
        rospy.loginfo("%s: Executing, creating fibonacci sequece of order %i with seeds %i, %i" % (
            self._action_name, goal.order, self._feedback.sequence[0], self._feedback.sequence[1]))
        
        # start executing the action
        for i in range(1, goal.order):
            # check that preempt has not been requested by the client
            if self._action_server.is_preempt_requested():
                rospy.loginfo("%s: Preempted" % self._action_name)
                self._action_server.set_preempted()
                success = False
                break
            self._feedback.sequence.append(self._feedback[i] + self._feedback.sequence)

            # publish the feedback
            self._action_server.publish_feedback(self._feedback)

            #this step is not necessary, the sequece is computed at 1Hz for demonstration purposes
            rate.sleep()

        if success:
            self._result.sequence = self._feedback.sequence
            rospy.loginfo("%s: Succeeded"% self._action_name)
            self._action_server.set_succeeded(self._result)

if __name__ == "__main__":
    rospy.init_node("fibonacii")
    server = FibonacciAction(rospy.get_name())
    rospy.spin()
    